### PR TITLE
3033 visualization flow

### LINKED ISSUE #3033: Data / Review visualization flow
- Start customizing a visualization
- Save it
- Load it
- Edit it after load it

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes #3033

## :v: What does this PR do?

The Visualization tab now has two views, one that shows the list, and the other that shows the Visualization in detail. In this view, you can modify and save the Visualization.

## :mag: How should this be manually tested?

Staging

### After this PR

3033_after

### CODE REVIEW

**Comment by Crashillo** [comment]
Solve conflicts in order to review

**Comment by jorgeatgu** [comment]
> Solve conflicts in order to review

Done!

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
-        class="gobierto-data-sql-editor-container-save-text"
-        :class="{
-          'disable-input-text': disabledButton,
-          'disable-cursor-pointer': enabledForkButton || !isUserLogged
-        }"
-        @keydown.stop="onKeyDownTextHandler"
-        @click="inputHandler"
-      >
+    <template v-if="isUserLogged">
+      <template v-if="isQuerySavingPromptVisible || labelValue ||isVizSavingPromptVisible">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Merge conditions

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 2]
```diff
@@ -1,25 +1,27 @@
...
+          @click="enabledInputHandler"
+        >
+      </template>
     </template>
 
     <!-- only show checkbox on prompt visible -->
-    <template v-if="isForkPromptVisible">
-      <template v-if="isSavingPromptVisible">
+    <template v-if="isForkPromptVisible && isUserLogged">
+      <template v-if="isQuerySavingPromptVisible || isVizSavingPromptVisible">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Same

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 3]
```diff
@@ -33,9 +35,12 @@
...
         </label>
       </template>
+    </template>
 
-      <!-- only show if label name is set OR the prompt is visible -->
-      <template v-if="isSavingPromptVisible || labelValue">
+
+    <!-- only show if label name is set OR the prompt is visible -->
+    <template v-if="isUserLogged">
+      <template v-if="isQuerySavingPromptVisible || labelValue ||isVizSavingPromptVisible">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Same

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 4]
```diff
@@ -46,75 +51,65 @@
...
-        class="btn-sql-editor"
-        @click.native="onClickSaveHandler"
-      />
+    <!-- show save button if there's no prompt but some name, otherwise, save button -->
+    <Button
+      v-if="!showForkButton"
+      :text="labelSave"
+      :disabled="!isDisabled"
+      icon="save"
+      color="var(--color-base)"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Didn't color come by default?

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 5]
```diff
@@ -191,12 +214,22 @@ export default {
...
   },
+  computed: {
+    isDisabled() {
+      return this.enabledVizSavedButton || this.enabledQuerySavedButton ? true : false
+    },
+    isDisabledFork() {
+      return this.enabledQuerySavedButton || this.isVizModified ? true : false
+    },
+    showForkButton() {
+      return this.enabledForkVizButton || this.enabledForkButton ? true : false
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: You may remove the `true : false`

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    if (userId) {
+      await this.getDataVisualization(this.privateVisualizations);
+      await this.getDataVisualization(this.publicVisualizations);
+    } else {
+      await this.getDataVisualization(this.publicVisualizations);
+    }
+  },
+  beforeDestroy() {
+    this.$root.$emit('disableVizModified')
+    this.$root.$emit('disabledSavedVizString')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Explain why do this on destroy

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+      }
+    },
+    updateVizName(value) {
+      const {
+        name: vizName
+      } = value;
+      this.labelValue = vizName
+      this.$root.$emit('isVizModified')
+      this.$root.$emit('enableSavedVizButton')
+      this.$root.$emit('disabledSavedVizString')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Why are there three events to update a name?

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    },
+    showSavingDialog() {
+      const userId = getUserId()
+      this.showVisualize = false
+      this.showResetViz = true
+      this.$root.$emit('enableSavedVizButton')
+      this.$root.$emit('disabledSavedVizString')
+      if (this.isUserLogged) {
+        this.$root.$emit("isVizSavingPromptVisible", true);
+        this.$root.$emit("isVizModified");
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: 4 events in a row, even with params. What for?

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsList.vue [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    };
+  },
+  methods: {
+    loadViz(vizName, user) {
+      const userId = Number(getUserId())
+      this.$emit('changeViz', 1)
+      this.$root.$emit('loadVizName', vizName)
+      if (userId !== 0 && userId !== user) {
+        this.$root.$emit('enabledForkVizButton', true)
+      }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsList.vue:179]: Do you like to emit events, right?

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 10]
```diff
@@ -227,13 +238,44 @@ export default {
...
+        this.saveHandlerNewQuery(opts)
+      }
+    },
+    saveHandlerSavedQuery(opts) {
       // send the query to be stored
       this.$root.$emit("storeCurrentQuery", opts);
+
+      this.$root.$emit('disabledSavedButton')
+      this.$root.$emit("resetToInitialState");
+      this.$root.$emit("isQuerySavingPromptVisible", false);
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: Yeah, you love it.

Why the save handler has to know that:
- there's a disabled button
- something has to reset
- and even, the saving prompt, has to be closed

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 11]
```diff
@@ -257,6 +299,22 @@ export default {
       //Avoid errors when user goes to the same route
       // eslint-disable-next-line no-unused-vars
       ).catch(err => {})
+    },
+    enabledInputQueries() {
+      const userId = getUserId();
+        if (!this.enabledForkButton && !!userId) {
+        this.$root.$emit('enableSavedButton')
+        this.$root.$emit('enabledForkPrompt')
+        this.$root.$emit('disabledRevertButton')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: 😭

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 12]
```diff
@@ -127,54 +122,66 @@ export default {
...
       this.perspectiveChanged = false
-      this.removeLabelBtn = false
       this.showResetViz = false
       this.typeChart = 'hypergrid'
-      this.isVisualizationModified = false
 
       this.$refs.viewer.enableDisabledPerspective(hidePerspective);
       this.$refs.viewer.setColumns();
+      this.$root.$emit('disabledSavedVizButton')
+      this.$root.$emit('disabledSavedVizString')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: 😭

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 13]
```diff
@@ -303,11 +356,19 @@ export default {
...
     this.$root.$off('disabledRevertButton')
+    this.$root.$off('isQuerySavingPromptVisible')
+    this.$root.$off('isVizSavingPromptVisible')
+    this.$root.$off('enableSavedVizButton')
+    this.$root.$off('disabledSavedVizString')
+    this.$root.$off('isVizModified')
+    this.$root.$off('loadVizName')
+    this.$root.$off('reloadVisualizations')
+    this.$root.$off('disableVizModified')
+    this.$root.$off('enabledForkVizButton')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Keep reading comments

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 14]
```diff
@@ -491,11 +627,28 @@ export default {
...
+
+      // Get the id if the query matches with a stored query
+      const { id } =
+        this.privateQueries.find(
+          ({ attributes: { sql } }) => sql === this.currentQuery
+        ) || {};
+
+      let currentQueryViz;
+
+      if (user !== userId) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:685]: This inequality means you're trying to save a query from another user?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 14 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue:685]: Yes, in this case, the user is trying to save a Viz from another user

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue (hunk 10 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: I hate them! :(

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 15]
```diff
@@ -1,25 +1,27 @@
...
+          @click="enabledInputHandler"
+        >
+      </template>
     </template>
 
     <!-- only show checkbox on prompt visible -->
-    <template v-if="isForkPromptVisible">
-      <template v-if="isSavingPromptVisible">
+    <template v-if="enableForkPrompt()">
+      <template v-if="handlerInputQuery()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Merge

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 16]
```diff
@@ -1,25 +1,27 @@
...
-        class="gobierto-data-sql-editor-container-save-text"
-        :class="{
-          'disable-input-text': disabledButton,
-          'disable-cursor-pointer': enabledForkButton || !isUserLogged
-        }"
-        @keydown.stop="onKeyDownTextHandler"
-        @click="inputHandler"
-      >
+    <template v-if="isUserLogged">
+      <template v-if="handlerInputQuery()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Merge

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 17]
```diff
@@ -33,9 +35,12 @@
...
         </label>
       </template>
+    </template>
+
 
-      <!-- only show if label name is set OR the prompt is visible -->
-      <template v-if="isSavingPromptVisible || labelValue">
+    <!-- only show if label name is set OR the prompt is visible -->
+    <template v-if="isUserLogged">
+      <template v-if="handlerInputQuery()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Merge

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 18]
```diff
@@ -294,16 +304,8 @@ export default {
...
-      const userId = getUserId();
-        if (!this.enabledForkButton && !!userId) {
-        this.disabledButton = false
-        this.$root.$emit('enableSavedButton')
-        this.$root.$emit('enabledForkPrompt')
-        this.$root.$emit('disabledRevertButton')
-        this.$nextTick(() => this.$refs.inputText.focus());
-      }
+    enabledInputHandler() {
+      this.$emit('enabledInput')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue:300]: Bravo!

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 19]
```diff
@@ -227,14 +238,41 @@ export default {
...
+        this.saveHandlerSavedQuery(opts)
+      } else {
+        this.saveHandlerNewQuery(opts)
+      }
+    },
+    saveHandlerSavedQuery(opts) {
       // send the query to be stored
       this.$root.$emit("storeCurrentQuery", opts);
     },
+    saveHandlerNewQuery(opts) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: I don't like these methods: `onSaveEventHandler`, `saveHandlerSavedQuery`, `saveHandlerNewQuery`, what does SQLEditorHeader care about whether is a new or saved query? Me (SQLEditorHeader) emits: "hey, I'm saving the query". That's all.
My listener (Dataset, for sure) listens: "Are you saving? OK, that means the querySavingPrompt is visible, etc, etc..."

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 20]
```diff
@@ -257,6 +295,20 @@ export default {
       //Avoid errors when user goes to the same route
       // eslint-disable-next-line no-unused-vars
       ).catch(err => {})
+    },
+    enabledInputQueries() {
+      const userId = getUserId();
+        if (!this.enabledForkButton && !!userId) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: Indentation

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 21]
```diff
@@ -127,54 +122,66 @@ export default {
...
   },
   methods: {
     onSaveEventHandler(opts) {
       // get children configuration
       const config = this.$refs.viewer.getConfig()
 
-      this.$root.$emit("storeCurrentVisualization", config, opts);
-      this.removeLabelBtn = true
-      this.isVisualizationModified = false
+      if (!this.isVizSavingPromptVisible) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: Same logic as I said before

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 22]
```diff
@@ -127,54 +122,66 @@ export default {
...
       this.showResetViz = true
-      this.isVisualizationModified = true
-      this.$root.$emit('enableSavedButton')
-      this.$root.$emit("isSavingPromptVisible", true);
-    },
-    resetButtonViz() {
-      this.removeLabelBtn = false
-      this.showResetViz = true
+      this.$root.$emit('enableSavedVizButton', true)
+      if (this.isUserLogged) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: For instance, what does _SQLEditorResults_ care about if the user is or not logged? It's not related

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 23]
```diff
@@ -0,0 +1,228 @@
+<template>
+  <div class="gobierto-data-sql-editor">
+    <template v-if="privateVisualizations.length && items || publicVisualizations.length && items">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Be careful with the ANDs & ORs, wrap the clauses with parenthesis

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 24]
```diff
@@ -0,0 +1,228 @@
...
+            :is-query-saving-prompt-visible="isQuerySavingPromptVisible"
+            @save="onSaveEventHandler"
+            @keyDownInput="updateVizName"
+            @handlerFork="handlerForkViz"
+          />
+          <Button
+            :text="labelEdit"
+            class="btn-sql-editor"
+            icon="chart-area"
+            color="var(--color-base)"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Color?

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+  methods: {
+    onSaveEventHandler(opts) {
+      //Add visualization ID to opts object, we need it to update a viz saved
+      opts.vizID = Number(this.vizID)
+      opts.user = Number(this.user)
+      opts.queryViz = Number(this.queryViz)
+      // get children configuration
+      const config = this.$refs.viewer.getConfig()
+
+      if (!this.isVizSavingPromptVisible) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: This is the third time (yep, I don't review in order) I see this snippet (you know what that means). But, does it require actually so much complexity just to _save_?

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+      this.$refs.viewer.enableDisabledPerspective(showPerspective);
+    },
+    showSavingDialog() {
+      const userId = getUserId()
+      this.showVisualize = false
+      this.showResetViz = true
+      //Enable saved button
+      this.$root.$emit('enableSavedVizButton', true)
+      //If user is logged show a saved button and string 'query modified'
+      if (this.isUserLogged) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Same about why the component need to know if the user is logged

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 27]
```diff
@@ -63,6 +67,16 @@
       v-else-if="activeDatasetTab === 3"
       :dataset-id="datasetId"
       :is-user-logged="isUserLogged"
+      :is-viz-saving-prompt-visible="isVizSavingPromptVisible"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:72]: Repeated

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 28]
```diff
@@ -63,6 +67,16 @@
...
+      :is-viz-modified="isVizModified"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:73]: Repeated

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 29]
```diff
@@ -63,6 +67,16 @@
...
+      :is-viz-saved="isVizSaved"
+      :is-private-viz-loading="isPrivateVizLoading"
+      :is-public-viz-loading="isPublicVizLoading"
+      :public-visualizations="publicVisualizations"
+      :private-visualizations="privateVisualizations"
+      :enabled-viz-saved-button="enabledVizSavedButton"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:79]: My god. Repeated. Use a linter or something.

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 30]
```diff
@@ -507,32 +663,39 @@ export default {
...
+        this.isVizModified = false
+        this.isVizSaved = true
+        this.isVizSavingPromptVisible = false
+        this.enabledVizSavedButton = false
+
+        await this.getPrivateVisualizations()
+        await this.getPublicVisualizations()
+
+        // TODO: indicar algo con el status OK
+        console.log("postVisualization", status);
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: You may remove this

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue (hunk 22 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: `SQLEditorResults` contains the `SavingDialog` of Visualizations. If the user is not logged in, we'll redirect him to the login.

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 31]
```diff
@@ -1,120 +1,108 @@
...
+
 
-      <!-- only show if label name is set OR the prompt is visible -->
-      <template v-if="isSavingPromptVisible || labelValue">
-        <PrivateIcon
-          :is-closed="isPrivate"
-          :style="{ paddingRight: '.5em', margin: 0 }"
-        />
+    <!-- only show if label name is set OR the prompt is visible -->
+    <template v-if="handlerInputQuery()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Avoid function calls in the DOM, unless they're belong to events (like `@click`). That creates multiple re-renders undesiderable. It's a bad practice.

In this case:
```suggestion
    <template v-if="handlerInputQuery">
```
And move the function `handlerInputQuery` to computed block, in order to be cached.

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 32]
```diff
@@ -1,120 +1,108 @@
...
+        >
     </template>
 
     <!-- only show checkbox on prompt visible -->
-    <template v-if="isForkPromptVisible">
-      <template v-if="isSavingPromptVisible">
-        <label
-          :for="labelPrivate"
-          class="gobierto-data-sql-editor-container-save-label"
+    <template v-if="enableForkPrompt()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: The same about the function calls in the DOM

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 33]
```diff
@@ -1,120 +1,108 @@
...
-        type="text"
-        class="gobierto-data-sql-editor-container-save-text"
-        :class="{
-          'disable-input-text': disabledButton,
-          'disable-cursor-pointer': enabledForkButton || !isUserLogged
-        }"
-        @keydown.stop="onKeyDownTextHandler"
-        @click="inputHandler"
-      >
+    <template v-if="handlerInputQuery()">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: And the same

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 34]
```diff
@@ -1,120 +1,108 @@
...
+      @click.native="onClickSaveHandler"
+    />
+
+    <Button
+      v-if="showForkButton"
+      :text="labelFork"
+      :title="labelButtonFork"
+      :disabled="!isDisabledFork"
+      icon="code-branch"
+      color="var(--color-base)"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Is necessary color?

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 35]
```diff
@@ -217,49 +243,31 @@ export default {
     }
   },
   methods: {
+    handlerInputQuery() {
+       return this.isUserLogged && this.isQuerySavingPromptVisible || this.labelValue || this.isVizSavingPromptVisible
+    },
+    enableForkPrompt() {
+      return this.isForkPromptVisible && this.isUserLogged && this.isQuerySavingPromptVisible || this.labelValue || this.isVizSavingPromptVisible
+    },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: To computed. And group the clauses

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 36]
```diff
@@ -0,0 +1,223 @@
+<template>
+  <div class="gobierto-data-sql-editor">
+    <template v-if="(privateVisualizations.length && items) || (publicVisualizations.length && items)">
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: when you think the logic becomes complex, a computed property will solve that out.

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 37]
```diff
@@ -0,0 +1,223 @@
...
+      this.showVisualize = false
+      this.showResetViz = true
+      //Enable saved button
+      this.$root.$emit('enableSavedVizButton', true)
+      this.$root.$emit("isVizModified", true);
+      this.$nextTick(() => this.$refs.savingDialogVizElement.inputFocus())
+
+      if (userId !== this.user) {
+        this.isQuerySavingPromptVisible = true
+      }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue]: Has `VisualizationItem` the decision the handle the user stuff, or perhaps their parents?

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsTab.vue [hunk 38]
```diff
@@ -133,94 +47,93 @@ export default {
...
-        // Append the visualization configuration
-        const visualization = { queryData, config: spec, name, privacy_status, query_id, id };
+    const {
+      name: nameComponent
+    } = this.$route;
 
-        visualizations.push(visualization);
-      }
+    if (nameComponent === 'Visualization') {
+      this.showVizElement(1)
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsTab.vue]: 👀

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 39]
```diff
@@ -257,6 +282,20 @@ export default {
       //Avoid errors when user goes to the same route
       // eslint-disable-next-line no-unused-vars
       ).catch(err => {})
+    },
+    enabledInputQueries() {
+      const userId = getUserId();
+      if (!this.enabledForkButton && !!userId) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: Why is the user necessary?

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 40]
```diff
@@ -16,31 +16,26 @@
...
+          :label-saved="labelSavedVisualization"
+          :label-modified="labelModifiedVizualition"
+          :is-viz-saving-prompt-visible="isVizSavingPromptVisible"
+          :is-viz-modified="isVizModified"
+          :is-viz-saved="isVizSaved"
+          :is-user-logged="isUserLogged"
+          :enabled-viz-saved-button="enabledVizSavedButton"
           @save="onSaveEventHandler"
-          @resetButtonViz="resetButtonViz"
+          @keyDownInput="updateVizName"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: Add `handler` to the event functions

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 41]
```diff
@@ -127,54 +126,60 @@ export default {
...
       this.perspectiveChanged = false
-      this.removeLabelBtn = false
       this.showResetViz = false
       this.typeChart = 'hypergrid'
-      this.isVisualizationModified = false
 
       this.$refs.viewer.enableDisabledPerspective(hidePerspective);
       this.$refs.viewer.setColumns();
+      this.$root.$emit('enableSavedVizButton', true)
+      this.$root.$emit('disabledSavedVizString')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: That's why I insist on this kind of events: you're resetting a visualization, why do you enable a button and disabled a string? why a function called `resetViz` has to know this sort of things?

The workflow should be the following: you emit "reset the viz", whoever is listening such event KNOWS that: the saving button must be enabled, AND the string disabled.

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 42]
```diff
@@ -127,54 +126,60 @@ export default {
...
       this.showResetViz = true
-      this.isVisualizationModified = true
-      this.$root.$emit('enableSavedButton')
-      this.$root.$emit("isSavingPromptVisible", true);
-    },
-    resetButtonViz() {
-      this.removeLabelBtn = false
-      this.showResetViz = true
+      this.$root.$emit('enableSavedVizButton', true)
+      this.$root.$emit("isVizModified", true);
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: Same explanation as before

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 43]
```diff
@@ -270,26 +311,43 @@ export default {
...
     // Reset to the default query
     this.$root.$on('resetQuery', this.resetQuery)
-    //reset to the saved query
+    // reset to the saved query
     this.$root.$on('revertSavedQuery', this.revertSavedQuery)
-
+    // activate button to saved a query
     this.$root.$on('enableSavedButton', this.activatedSavedButton)
-
+    // disable
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: ejem ejem

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue (hunk 39 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: It's unnecessary. It's enough to check if the user is logged.

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 44]
```diff
@@ -491,11 +658,35 @@ export default {
...
       }
     },
     async storeCurrentVisualization(config, opts) {
 
-      const { name, privacy } = opts;
+      if (!this.isVizSavingPromptVisible) {
+        this.isVizSavingPromptVisible = true
+        this.vizInputFocus = true
+      } else if (!this.vizName) {
+        this.vizInputFocus = true
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:670]: is this ok? Get it out from the conditional clause if it's the same value

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 45]
```diff
@@ -1,120 +1,106 @@
...
+    <template v-if="handlerInputQuery">
+        <input
+          ref="inputText"
+          v-model="labelValue"
+          :placeholder="placeholder"
+          type="text"
+          class="gobierto-data-sql-editor-container-save-text"
+          :class="{
+            'disable-input-text': disabledButton,
+            'disable-cursor-pointer': enabledForkButton || !isUserLogged || enabledForkVizButton
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue:13]: If you see the conditionals getting bigger, you may create a computed prop

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 46]
```diff
@@ -191,12 +205,28 @@ export default {
...
+      return this.enabledQuerySavedButton || this.isVizModified
+    },
+    showForkButton() {
+      return this.enabledForkVizButton || this.enabledForkButton
+    },
+    handlerInputQuery() {
+       return (this.isUserLogged && this.isQuerySavingPromptVisible) || this.labelValue || this.isVizSavingPromptVisible
+    },
+    enableForkPrompt() {
+      return (this.isForkPromptVisible && this.isUserLogged && this.isQuerySavingPromptVisible) || this.labelValue || this.isVizSavingPromptVisible
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue:227]: Something similar even here, check this out https://vuejs.org/v2/style-guide/#Simple-computed-properties-strongly-recommended

File: app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue [hunk 47]
```diff
@@ -0,0 +1,223 @@
...
+    }
+  },
+  async created() {
+    const userId = getUserId()
+    if (userId) {
+      await this.getDataVisualization(this.privateVisualizations);
+      await this.getDataVisualization(this.publicVisualizations);
+    } else {
+      await this.getDataVisualization(this.publicVisualizations);
+    }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/VisualizationsItem.vue:153]: You could simplify this, applying the conditional only to the private clause

**Review by Crashillo — APPROVED** [review]

**Pull request merged**